jwst/emicorr/emicorr.py

```diff
@@ -502,203 +644,594 @@ def sloper(data):
+        for a single amplitude and phase across all integrations.
+    nphases_opt : int, optional
+        Number of phases to sample chi squared as a function of phase
 
     Returns
     -------
-    Nothing
+    data : ndarray
+        The input 4D data array with the model waveform phased, scaled,
+        and subtracted.
     """
-    # save the reference file if save_intermediate_results=True and no CRDS file exists
-    emicorr_model = datamodels.EmiModel(freq_pa_dict)
-    if 'fits' in emicorr_ref_filename:
-        emicorr_ref_filename = emicorr_ref_filename.replace('fits', 'asdf')
-    emicorr_model.save(emicorr_ref_filename)
-    emicorr_model.close()
-    log.info('On-the-fly reference file written as: %s', emicorr_ref_filename)
+    nints, ngroups, ny, nx = data.shape
+
+    # divide ncols by the four output channels (which are read simultaneously)
+    nx4 = nx // 4
+
+    # non-roi rowclocks between subarray frames (this will be 0 for fullframe)
+    extra_rowclocks = (1024.0 - ny) * (4 + 3.0)
+
+    frametime = ny * rowclocks + extra_rowclocks
+
+    # Times of the individual reads in pixel clock times
+    t0_arr = np.zeros((ny, nx4))
+    for i in range(ny):
+        t0_arr[i] = i * rowclocks + np.arange(nx4) * nsamples
+    phase = (t0_arr / period_in_pixels) % 1
+
+    # Phase gap between groups
+    dphase = (frametime / period_in_pixels) % 1
+
+    # Phase gap between integrations
+    dphase_frame = dphase * ngroups + frameclocks / period_in_pixels
+
+    # Time-like index for group number
+    grouptimes = np.arange(ngroups)
+
+    # Phase(model) for interpolation.  Wrap the phases to avoid
+    # extrapolation, and use a cubic spline.
+
+    refwave_extended = np.array([refwave[-1]] + list(refwave) + [refwave[0]])
+    phase_extended = (np.arange(len(refwave) + 2) - 0.5) / len(refwave)
+    phasefunc = interpolate.interp1d(phase_extended, refwave_extended, kind="cubic")
+
+    phases_template = (phase_extended[1:-1, np.newaxis] + grouptimes * dphase) % 1
+    nphases = phases_template.shape[0]
+
+    # These arrays hold the sum of the values of good pixels at a
+    # given phase and the total number of good pixels at a given
+    # phase, respectively.  The phase refers to the first group; other
+    # groups will have the appropriate phase delay added.
+
+    all_y = np.zeros((nints, nphases, ngroups))
+    all_n = np.zeros((nints, nphases))
+
+    # "Good" pixel here has no more than twice the median standard
+    # deviation among group differences and is not flagged in the pdq
+    # array.  This should discard most bad and high-flux pixels.
+
+    pixel_std = np.std(data, axis=1)
+    pixel_ok = (pixel_std < 2 * np.median(pixel_std)) & (pdq == 0)
+
+    for j in range(ny):
+        for k in range(nx4):
+            # Bad reference column?  I am not sure whether this is needed,
+            # but I think so. I carried it from the current routine.
```

Comment:
I think we can just say "bad reference column" here--I don't think we need the following qualifier.